First version of a generator for Nix expressions to build packages

Ma27 opened the pull request:
This PR addresses the second feature request of #23 

This is just a first basic version to review (things tested against a test project locally ATM), the following things are missing, before this is mergable

- [x] first generator command
- [x] use SBT parameter to make `--with-composition` obsolete
- [x] option for `buildSbtProject` and `buildSbtLibrary`
  - [x] `buildSbtLibrary`
  - [x] `buildSbtPackage`
  - [x] deduplicate expressions
- [x] tests, tests, tests!
- [x] check IDE integration
- [x] docs

Ma27 committed:
First version of a generator for Nix expressions to build packages

Ma27 committed:
Get rid of `--with-composition` option; prefer build parameters from SBT

teozkr commented:
Looks great :D 

Quick nitpick, but please put resources in `src/main/resources` instead of changing `resourceDirectory`.

Ma27 committed:
Avoid explicit configuration of `resourceDirectory`

Ma27 commented:
done. Will do some further progress in the next days :)

Ma27 committed:
Refactor nix expression

- use 2 spaces for indent to comply with NixOS codestyle
- don't leak `sbtixDir` into `sbtix.build` expression

Ma27 committed:
Make composition type (e.g. `library` / `program`) configurable through SBT

Ma27 committed:
Don't override existing composition files

Ma27 committed:
Support `sbtix.buildSbtProject` type

Ma27 commented:
@teozkr I got a concept question ATM and I'd like to discuss first before I start delivering more code:
ATM we have three(!) nix expressions (for `library`, `project` and `program`). Except some minor details they are almost identical which is bad as we would have to maintain several files now in case of API changes in our nix-layer.

One idea would be to use something like `twirl` as it's IMO not a good idea to add even more `.replace()` statements to the `genComposition` command. Furthermore I wanted to generate a skeleton of an `installPhase` when `buildSbtProject` is used, so we would need some sort of `if`-condition here. An idea would be to use something like `twirl` to delegate this sort of logic into a template, but I'm not 100% confident because this would add tons of new dependencies here and would be inconsistent to the way, the `repo.nix` is generated (I don't think we need two completely different implementations for an almost equal problem).
Another idea would be to add another object named `CompositionWriter` which generates the `default.nix` and is somehow consistent to the `NixWriter` as it builds the expression without another templating language.

Ma27 committed:
Drop templates with duplicated content in favor of a simple code generator

Ma27 commented:
I just thought a little bit about IDE integration: the only issue I can image is that some `build.sbt` expressions are not found, but this can be fixed by using this via `addSbtPlugin`.
In the end I don't think we can do too much about it here as IntelliJ uses (AFAIK) it's own SBT setup for building from an IDE.

Ma27 committed:
Implement proper testing for the composition generator

Ma27 committed:
Add some more information about the NixOS compositions in the docs

teozkr commented:
Sorry about the very slow replies, I haven't been at my computer much the last couple of days. 

Anyway, I'd prefer a template approach here, since IMO we're not gaining much from a full Nix AST and it would complicate adding template comments, etc. Freezing the repo is different, since we're just treating it as glorified JSON there. In fact, perhaps we should just move to that and drop the AST completely? 

I'm not sure we should add a generator for buildSbtProject, since that's just an implementation detail (and escape hatch) that generally shouldn't be needed directly. 

Finally, I'm not sure it's a good idea to add sbt settings since projects should still be buildable without the Sbtix wrapper.  That said, it also feels ugly to add our own settings mechanism..  :/

Ma27 commented:
thanks for your response.

> Anyway, I'd prefer a template approach here, since IMO we're not gaining much from a full Nix AST and it would complicate adding template comments, etc. Freezing the repo is different, since we're just treating it as glorified JSON there. In fact, perhaps we should just move to that and drop the AST completely?

The reason why I added yet another generator is quite easy: the generation of the `default.nix` file doesn't contain much divergent logic, so I don't get why we should add such an overhead. In this case I simply modified a string and removed all trailing spaces, so I'm unsure if we really need a templating engine here.

> I'm not sure we should add a generator for buildSbtProject, since that's just an implementation detail (and escape hatch) that generally shouldn't be needed directly.

So you mean that we only support `buildSbtProgram` and `buildSbtLibrary` and throw a warning when `buildSbtProject` should be used? (I simply added it as it was mentioned in the docs as more generic solution)

> Finally, I'm not sure it's a good idea to add sbt settings since projects should still be buildable without the Sbtix wrapper. That said, it also feels ugly to add our own settings mechanism.. :/

I respectfully disagree here because of the following reasons:

- we already have several configuration parameters in `build.sbt` that are only changeable when adding them to our SBT build (I don't think that we want to add another configuration mechanism, the current one is quite straight-forward and should be kept because of consistency)

- why not adding `sbtix` as plugin in your `project/plugins.sbt`? Many projects use such an approach: an example would be the `sbt` plugin of `scalastyle`: it can be added as plugin and then you can modify several things in your `build.sbt` and have new commands available when running `sbt ...` on your command line. I think that it's a better idea to recommend such an approach in our docs as it doesn't force the devs to add another package to their nix-profile, but they can run all relevant actions in their projects.

teozkr commented:
> The reason why I added yet another generator is quite easy: the generation of the default.nix file doesn't contain much divergent logic, so I don't get why we should add such an overhead. In this case I simply modified a string and removed all trailing spaces, so I'm unsure if we really need a templating engine here.

Sounds like a simple template to me. :P What I meant is that we probably don't want to go the ScalaTags approach, for now at least.

But yeah, I hadn't read your code yet and it's doing the right thing. My bad.

> So you mean that we only support buildSbtProgram and buildSbtLibrary and throw a warning when buildSbtProject should be used? (I simply added it as it was mentioned in the docs as more generic solution)

Not sure what you mean here? I'm mostly talking about removing the documentation for that case, and not having the special-case installPhase.

> * we already have several configuration parameters in build.sbt that are only changeable when adding them to our SBT build (I don't think that we want to add another configuration mechanism, the current one is quite straight-forward and should be kept because of consistency)

The only setting that a normal build would be interested in changing is `nixRepoFile`, which has the same problem. However, that has a sensible default value, where the project type is something that ~50% of projects would need to flip no matter what default we choose.

I agree that it's ugly to add a new config system on top of SBT's, but I'm not sure what other options we have.

> * why not adding sbtix as plugin in your project/plugins.sbt? Many projects use such an approach: an example would be the sbt plugin of scalastyle: it can be added as plugin and then you can modify several things in your build.sbt and have new commands available when running sbt ... on your command line. I think that it's a better idea to recommend such an approach in our docs as it doesn't force the devs to add another package to their nix-profile, but they can run all relevant actions in their projects.

You don't need to add it to your nix-profile, you can run it from a nix-shell as well.

But aside from that, being addable using addSbtPlugin requires publishing our plugin to Maven Central, which means we can't depend on Nix packages like `nix-prefetch-scripts` anymore.

That said, perhaps we could compromise, and publish an "interface" plugin to Maven Central that only contains the settings. That would mean that the settings could still be changed the SBT way, but genNix would require the Sbtix wrapper.

Ma27 commented:
> Not sure what you mean here? I'm mostly talking about removing the documentation for that case, and not having the special-case installPhase.

You mentioned that `buildSbtProject` is basically an implementation detail and the new generator I created shouldn't generate such an expression and shouldn't care about adding an `installPhase` expression, right?
I proposed that when a user wants to generate a `default.nix` with a `sbtix.buildSbtProject` call we should raise a warning (or even abort with an error).

> I agree that it's ugly to add a new config system on top of SBT's, but I'm not sure what other options we have.

I think that there's a second important reason why we want to configure it using SBT rather than using a `nix-shell` or building it into the `nix-profile`: I think that it might a long time until every developer is convinced to use Nix and I think that when a non-Nix user adds new dependencies it might be way easier if he could use his `sbt` installation to refresh the `repo.nix` file, so everyone can add new deps easily.

> But aside from that, being addable using addSbtPlugin requires publishing our plugin to Maven Central, which means we can't depend on Nix packages like nix-prefetch-scripts anymore.That said, perhaps we could compromise, and publish an "interface" plugin to Maven Central that only contains the settings. That would mean that the settings could still be changed the SBT way, but genNix would require the Sbtix wrapper

well, that's indeed a problem. I'll think about it later this day, ok? :-)

teozkr commented:
> You mentioned that buildSbtProject is basically an implementation detail and the new generator I created shouldn't generate such an expression and shouldn't care about adding an installPhase expression, right?
I proposed that when a user wants to generate a default.nix with a sbtix.buildSbtProject call we should raise a warning (or even abort with an error).

Alright, that sounds OK.

> I think that there's a second important reason why we want to configure it using SBT rather than using a nix-shell or building it into the nix-profile: I think that it might a long time until every developer is convinced to use Nix and I think that when a non-Nix user adds new dependencies it might be way easier if he could use his sbt installation to refresh the repo.nix file, so everyone can add new deps easily.

True. But then we couldn't import it into the user's Nix store from Coursier's cache. I suppose we could, however, have two paths: one where we calculate the hash ourselves, and one where we import into Nix, depending on what's available on the system.

Ma27 commented:
> I suppose we could, however, have two paths: one where we calculate the hash ourselves, and one where we import into Nix, depending on what's available on the system.

That sound's interesting, but I think that this should live in another PR, right?

IMO we should do the following things to finalize this PR:

- implement the warning when trying to use `buildSbtProject`
- find suitable solution for the plugin problem: we could either set more sensible defaults (no `default.nix` by default) or keep add another configuration layer on top (sth. like `SBTIX_COMPOSITION=1 sbtix-gen-all ...` for instance)

teozkr commented:
Sounds good. I'd say default conservatively then, since it'd be relatively pointless to add a new config mechanism just to tear it down again.

Ma27 commented:
ok, will finish the remaining tasks next week :-)

Ma27 committed:
Drop support for `sbtix.buildSbtProject` as it's part of the interal API

Ma27 committed:
Use more conservative defaults (no composition generation at first) to avoid clashes with non-Nix SBT setups

Ma27 commented:
@teozkr I finally managed to reserve some time to fix the remaining issues we talked about.
All other aspects should be covered by other issues I'll try to work on in the next time.

This PR is basically ready, so I'd love to get a review from your side :-)

teozkr commented:
LGTM, and Travis seems happy. Thanks!
